internals/overlord/logstate/gatherer.go

```diff
@@ -0,0 +1,207 @@
+}
+
+func (g *logGatherer) planChanged(pl *plan.Plan, buffers map[string]*servicelog.RingBuffer) {
+	g.pullersLock.Lock()
+	defer g.pullersLock.Unlock()
+
+	newPullers := map[string]*logPuller{}
+
+	for _, service := range pl.Services {
+		target := pl.LogTargets[g.targetName]
+		if !service.LogsTo(target) {
+			continue
+		}
+
+		if puller, ok := g.pullers[service.Name]; ok {
+			// Move over existing puller
+			newPullers[service.Name] = puller
+			delete(g.pullers, service.Name)
+			continue
+		}
+
+		buffer, bufferExists := buffers[service.Name]
+		if !bufferExists {
+			// We don't yet have a reference to the service's ring buffer
+			// Need to wait until serviceStarted
+			continue
+		}
+
+		// Create new puller
+		newPullers[service.Name] = newLogPuller(g.entryCh, buffer)
+	}
+
+	// Old pullers for now-removed services need to be shut down.
+	for svc, puller := range g.pullers {
+		err := puller.tomb.Killf("gatherer closed")
+		if err != nil {
+			logger.Noticef("Error shutting down puller for target %q, service %q: %v",
+				g.targetName, svc, err)
+		}
+	}
+	g.pullers = newPullers
+}
+
+func (g *logGatherer) serviceStarted(service *plan.Service, buffer *servicelog.RingBuffer) {
+	g.pullersLock.Lock()
+	defer g.pullersLock.Unlock()
+	g.pullers[service.Name] = newLogPuller(g.entryCh, buffer)
+}
+
+func (g *logGatherer) loop() {
+	ticker := time.NewTicker(g.tickPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			// Timeout - flush
+			err := g.client.Flush()
+			if err != nil {
+				logger.Noticef("Cannot send logs to target %q: %v", g.targetName, err)
+			}
+
+		case <-g.cancel:
+			// Gatherer has been stopped - flush any remaining logs
+			err := g.client.Flush()
+			if err != nil {
+				logger.Noticef("Cannot send logs to target %q: %v", g.targetName, err)
+			}
+			return
+
+		case entry := <-g.entryCh:
+			err := g.client.Write(entry)
+			if err != nil {
+				logger.Noticef("Cannot write logs to target %q: %v", g.targetName, err)
+			}
+		}
+	}
+}
+
+// stop closes the cancel channel, thereby terminating the main loop.
+func (g *logGatherer) stop() {
+	close(g.cancel)
+
+	g.pullersLock.Lock()
+	defer g.pullersLock.Unlock()
+	for _, puller := range g.pullers {
+		_ = puller.tomb.Killf("gatherer closed")
+	}
+}
+
+// logPuller handles pulling logs from a single iterator and sending to the
+// main control loop.
+type logPuller struct {
+	iterator servicelog.Iterator
+	tomb     *tomb.Tomb
+	entryCh  chan servicelog.Entry
+}
+
+func newLogPuller(entryCh chan servicelog.Entry, buffer *servicelog.RingBuffer) *logPuller {
+	p := &logPuller{
+		iterator: buffer.TailIterator(),
+		tomb:     &tomb.Tomb{},
+		entryCh:  entryCh,
+	}
+	p.tomb.Go(p.loop)
+	return p
+}
+
+func (p *logPuller) loop() error {
+	parser := servicelog.NewParser(p.iterator, parserSize)
+
+	for p.iterator.Next(p.tomb.Dying()) {
+		for parser.Next() {
+			if err := parser.Err(); err != nil {
+				return err
+			}
+			p.entryCh <- parser.Entry()
```

Comment:
Should be a send in a select statement, with an case that handles cancellation via tomb dying channel.